File: leetcode/linked_lists/ruby/merge_two_sorted_lists.rb
=begin
https://leetcode.com/problems/merge-two-sorted-lists/

Merge two sorted linked lists and return it as a new list. The new list should be made by splicing together the nodes of the first two lists.

Example:

Input: 1->2->4, 1->3->4
Output: 1->1->2->3->4->4
=end

# Definition for singly-linked list.
class ListNode
  attr_accessor :val, :next

  def initialize(val)
    @val = val
    @next = nil
  end
end

def sorted_merge(a, b)
  # base cases
  return b if a.nil?
  return a if b.nil?

  if a.val <= b.val
    result =  a
    result.next = sorted_merge(a.next, b)
  else
    result = b
    result.next = sorted_merge(a, b.next)
  end

  result
end

a = ListNode.new(1)
b = ListNode.new(2)
c = ListNode.new(3)

e = ListNode.new(4)
f = ListNode.new(5)
g = ListNode.new(6)

a.next = b
b.next = c

e.next = f
f.next = g

res = sorted_merge(a, e)

until res.nil?
  puts res.val
  puts
  res = res.next
end

File: leetcode/linked_lists/python/merge_two_sorted_lists.py
from linked_list import *

# https://leetcode.com/problems/merge-two-sorted-lists/

def merge_two_lists(a, b):
    # base cases
    if a == None:
        return b
    elif b == None:
        return a

    # pick either a or b and recur
    if a.data <= b.data:
        result = a
        result.next = merge_two_lists(a.next, b)
    else:
        result = b
        result.next = merge_two_lists(a, b.next)

    return result

# https://discuss.leetcode.com/topic/21292/python-solutions-iteratively-recursively-iteratively-in-place
def merge_two_lists_1(a, b):
    dummy = current = Node(0)

    while a and b:
        if a.data < b.data:
            current.next = a
            a = a.next
        else:
            current.next = b
            b = b.next

        current = current.next

    current.next = a or b
    return dummy.next

if __name__ == '__main__':
    a = Node(1)
    b = Node(2)
    c = Node(3)


    e = Node(4)
    f = Node(5)
    g = Node(6)

    a.next = b
    b.next = c

    e.next = f
    f.next = g

    # 1 -> 2 -> 3
    # 4 -> 5 -> 6

    res = merge_two_lists(a, e)

    while res != None:
        print str(res.data) + "\n"
        res = res.next
